### PR TITLE
Feature: jax integration

### PR DESCRIPTION
**Pull request by mrava87**
## Motivation 
This PR introduces a new backend in PyLops to enable using JAX arrays. 

As a by-product of JAX-enabled operators, we inherit JAX features like jit, automatic differentiation, and automatic vectorization.

## Highlights

- Created new `JaxOperator`
- Modified most of the operators and methods in LinearOperator to enable JAX integration
- Cleaned up `backend` module with new logic to detect whether np,cp, or jnp methods should be used based on the input type
- Added a new tutorial named `jaxop`
- Revamped `gpu.rst` documentation page

### CODE REVIEW

File: pylops/avo/poststack.py [hunk 1]
```diff
@@ -93,12 +94,15 @@ def _PoststackLinearModelling(
             D = ncp.diag(0.5 * ncp.ones(nt0 - 1, dtype=dtype), k=1) - ncp.diag(
                 0.5 * ncp.ones(nt0 - 1, dtype=dtype), -1
             )
-            D[0] = D[-1] = 0
+            # D[0] = D[-1] = 0.
```
Comment (VascoSch92) [inline pylops/avo/poststack.py:97]: why this comment? :-)

File: pylops/jaxoperator.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+            Input array
+        y : :obj:`jaxlib.xla_extension.ArrayImpl`
+            Output array (where to store the
+            Vector-Jacobian product)
+
+        Returns
+        ----------
+        y : :obj:`jaxlib.xla_extension.ArrayImpl`
+            Output array
+
```
Comment (VascoSch92) [inline pylops/jaxoperator.py:62]:
```suggestion
```

File: pylops/jaxoperator.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        Returns
+        -------
+        x : :obj:`numpy.ndarray`
+            Output array of shape (N,) or (N,1)
+
+        """
+        M, N = self.shape
+
+        if x.shape != (M,) and x.shape != (M, 1):
+            raise ValueError("dimension mismatch")
```
Comment (VascoSch92) [inline pylops/jaxoperator.py:87]:
```suggestion
            raise ValueError(f"Dimension mismatch. Got {x.shape}, but expected {(M, 1)} or {(M,)}.")
```

File: pylops/jaxoperator.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+        y = self._rmatvecad(x, y)
+
+        if x.ndim == 1:
+            y = y.reshape(N)
+        elif x.ndim == 2:
+            y = y.reshape(N, 1)
+        else:
+            raise ValueError("invalid shape returned by user-defined rmatvecad()")
```
Comment (VascoSch92) [inline pylops/jaxoperator.py:96]:
```suggestion
            raise ValueError("Invalid shape returned by user-defined rmatvecad(). Expected 1, or 2, but got {x.ndim}")
```

File: pylops/signalprocessing/fredholm1.py [hunk 5]
```diff
@@ -61,7 +61,7 @@ class Fredholm1(LinearOperator):
         d(k, x, z) = \int{G(k, x, y) m(k, y, z) \,\mathrm{d}y}
         \quad \forall k=1,\ldots,n_{slice}
 
-    on the other hand its adjoint is expressed as
+    on the other hand its adjoin is expressed as
```
Comment (VascoSch92) [inline pylops/signalprocessing/fredholm1.py:64]: i think it was correct not?

**Review by VascoSch92 — COMMENTED** [review]
Hey, great job :-) 

I left just some nit ;-)

File: pylops/jaxoperator.py (hunk 2 above)
Comment (mrava87) [reply pylops/jaxoperator.py:62]: Wrong, fixed now

File: pylops/avo/poststack.py (hunk 1 above)
Comment (mrava87) [reply pylops/avo/poststack.py:97]: Forgot to remove it

File: pylops/jaxoperator.py (hunk 3 above)
Comment (mrava87) [reply pylops/jaxoperator.py:87]: Done

File: pylops/jaxoperator.py (hunk 4 above)
Comment (mrava87) [reply pylops/jaxoperator.py:96]: Done

File: pylops/signalprocessing/fredholm1.py (hunk 5 above)
Comment (mrava87) [reply pylops/signalprocessing/fredholm1.py:64]: adjoint is the right word

**Comment by mrava87** [comment]
> Hey, great job :-)
> 
> I left just some nit ;-)

Thanks!

File: pylops/signalprocessing/fredholm1.py (hunk 5 above)
Comment (VascoSch92) [reply pylops/signalprocessing/fredholm1.py:64]: yes but now is `adjoin` :-)

Comment (mrava87) [reply pylops/signalprocessing/fredholm1.py]: Ops sorry, I read the comment the other way.. fixed now

**Comment by mrava87** [comment]
@cako you may want to look at this as supporting document https://github.com/PyLops/pylops_notebooks/blob/master/developement-cupy/Timing_CupyJAX.ipynb. It contains timing for most of the methods ported to the Jax backend and a comparison with numpy and cupy

File: docs/source/conf.py [hunk 6]
```diff
@@ -21,6 +21,7 @@
     "numpydoc",
     "nbsphinx",
     "sphinx_gallery.gen_gallery",
+    "sphinxemoji.sphinxemoji",
```
Comment (cako) [inline docs/source/conf.py:24]: What is this for?

File: pylops/waveeqprocessing/wavedecomposition.py [hunk 7]
```diff
@@ -187,6 +188,9 @@ def _obliquity3D(
     composition : :obj:`bool`, optional
         Create obliquity factor for composition (``True``) or
         decomposition (``False``)
+    fftengine : :obj:`str`, optional
+        Engine used for fft computation (``numpy`` or ``scipy``). Choose
+        ``numpy`` when working with cupy and jax arrays.
```
Comment (cako) [inline pylops/waveeqprocessing/wavedecomposition.py:193]: This needs a "new in vX.X"

File: pylops/jaxoperator.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+from typing import Any, NewType
+
+from pylops import LinearOperator
+from pylops.utils import deps
+
+if deps.jax_enabled:
+    import jax
+    import jaxlib
+
+    jaxarray_type = jaxlib.xla_extension.ArrayImpl
```
Comment (cako) [inline pylops/jaxoperator.py:14]: My understanding is that jax should be typed with `jax.Array` or `jax.typing.ArrayLike`: https://jax.readthedocs.io/en/latest/jax.typing.html

File: pylops/jaxoperator.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+        """Vector-Jacobian product
+
+        JIT-compiled Vector-Jacobian product
+
+        Parameters
+        ----------
+        x : :obj:`jaxlib.xla_extension.ArrayImpl`
+            Input array for forward
+        y : :obj:`jaxlib.xla_extension.ArrayImpl`
+            Input array for adjoint
```
Comment (cako) [inline pylops/jaxoperator.py:76]: Same here (typing)

**Review by cako — APPROVED** [review]
@mrava87 nicely done! Going to leave it as approved, but please have a look at some of the comments and my commits.

By the way I ran the notebook... seems like Jax is generally slower than CuPy? Am I reading this wrong?

image

File: pylops/utils/dottest.py [hunk 10]
```diff
@@ -93,13 +93,13 @@ def dottest(
     # make u and v vectors
     rdtype = np.ones(1, Op.dtype).real.dtype
 
-    u = ncp.random.randn(nc).astype(rdtype)
+    u = randn(nc, backend=backend).astype(rdtype)
```
Comment (cako) [inline pylops/utils/dottest.py:96]: The backend argument is now keyword-only in `randn`

**Comment by mrava87** [comment]
> @mrava87 nicely done! Going to leave it as approved, but please have a look at some of the comments and my commits.
> 
> By the way I ran the notebook... seems like Jax is generally slower than CuPy? Am I reading this wrong?
> 
> image

This is also what I see when running this both locally and on colab... my guess/suspicion is that when the operator has a limited number of steps all calling np/cp, cupy is already very well optimized so the `jit` of jax does not really do much... and for some reason the equivalent `jax.numpy` methods are apparently slower... in one case, where the operator matvec/rmatvec has a for...loop (`NonStationaryConvolve1D`), then jax seems to shine...

I read a lot about jax being very optimized for GPUs/TPUs so this was also an exercise to compare it with cupy, but so far what I observe is somehow that cupy is better ;)

File: docs/source/conf.py (hunk 6 above)
Comment (mrava87) [reply docs/source/conf.py:24]: This is for emojiis like `:white_check_mark:` as used in the `gpu.rst` file

File: pylops/jaxoperator.py (hunk 8 above)
Comment (mrava87) [reply pylops/jaxoperator.py:14]: Sorry, not sure why I used that one... fixed

File: pylops/jaxoperator.py (hunk 9 above)
Comment (mrava87) [reply pylops/jaxoperator.py:76]: Fixed

File: pylops/utils/dottest.py (hunk 10 above)
Comment (mrava87) [reply pylops/utils/dottest.py:96]: Sounds good :)

**Pull request merged**